src/target/riscv/riscv.c

```diff
@@ -1589,21 +1590,100 @@ static int riscv_trigger_detect_hit_bits(struct target *target, int64_t *unique_
+				if ((tdata1 & hit_mask) == 0) {
+					uint64_t tdata1_test =
+					set_field(tdata1, CSR_MCONTROL_HIT, 1);
+					if (riscv_reg_set(target,
+					GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
+						return ERROR_FAIL;
+					if (riscv_reg_get(target,
+					&tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+						return ERROR_FAIL;
+					int tdata1_test_hit =
+					get_field(tdata1_test, CSR_MCONTROL_HIT);
+					int tdata1_test_rb_hit =
+					get_field(tdata1_test_rb, CSR_MCONTROL_HIT);
+					if (tdata1_test_hit != tdata1_test_rb_hit)
+						mcontrol_hit_not_supported = true;
+					if (riscv_reg_set(target, GDB_REGNO_TDATA1,
+						tdata1_test & ~hit_mask) != ERROR_OK)
+						return ERROR_FAIL;
+				}
+
+				if (get_field(tdata1, CSR_MCONTROL_TIMING)
+				== CSR_MCONTROL_TIMING_BEFORE
+				    && (mcontrol_hit_not_supported || (tdata1 & hit_mask)))
+					r->need_single_step = true;
 				break;
 			case CSR_TDATA1_TYPE_MCONTROL6:
 				hit_mask = CSR_MCONTROL6_HIT0 | CSR_MCONTROL6_HIT1;
+				int hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+				int hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+				int trigger_retired_info = (hit1 << 1) | hit0;
+				/* Check if mcontrol6.hit0[1] is not implemented. */
+				if (trigger_retired_info == 0) {
+					for (unsigned int j = 0; j < 2; j++) {
+						uint64_t tdata1_test =
+							set_field(tdata1, CSR_MCONTROL6_HIT0, j);
+						for (unsigned int k = 0; k < 2; k++) {
+							tdata1_test =
+							set_field(tdata1_test, CSR_MCONTROL6_HIT1, k);
+							if (riscv_reg_set(target,
+								GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
+								return ERROR_FAIL;
+							if (riscv_reg_get(target,
+								&tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+								return ERROR_FAIL;
+							int tdata1_test_hit0 =
+								get_field(tdata1_test, CSR_MCONTROL6_HIT0);
+							int tdata1_test_rb_hit0 =
+								get_field(tdata1_test_rb, CSR_MCONTROL6_HIT0);
+							int tdata1_test_hit1 =
+								get_field(tdata1_test, CSR_MCONTROL6_HIT1);
+							int tdata1_test_rb_hit1 =
+								get_field(tdata1_test_rb, CSR_MCONTROL6_HIT1);
+							int trigger_retired_test_info =
+								(tdata1_test_hit1 << 1) | tdata1_test_hit0;
+							int trigger_retired_test_rb_info =
+								(tdata1_test_rb_hit1 << 1) | tdata1_test_rb_hit0;
+							if (trigger_retired_test_info == 0)
+								continue;
+
+							if (trigger_retired_test_info
+								!= trigger_retired_test_rb_info) {
+								mcontrol6_hit_not_supported = true;
+							} else {
+								mcontrol6_hit_not_supported = false;
+								if (riscv_reg_set(target, GDB_REGNO_TDATA1,
+									tdata1_test & ~hit_mask) != ERROR_OK)
+									return ERROR_FAIL;
+								goto done;
+							}
+						}
+						if (riscv_reg_set(target, GDB_REGNO_TDATA1,
+							tdata1_test & ~hit_mask) != ERROR_OK)
+							return ERROR_FAIL;
+					}
+				}
+done:
+				if (mcontrol6_hit_not_supported
+				    || trigger_retired_info == CSR_MCONTROL6_HIT0_BEFORE)
+					r->need_single_step = true;
```

Comment:
`step` is not needed if the trigger was not hit or of the trigger was hit after.
Therefore this can be simplified:
```suggestion
				r->need_single_step = true;
				if (trigger_retired_info == CSR_MCONTROL6_HIT0_FALSE) {
					/* Check timing "before" support. */
					const riscv_reg_t tdata1_test = set_field(tdata1, CSR_MCONTROL6_HIT0, CSR_MCONTROL6_HIT0_BEFORE);
					if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
						return ERROR_FAIL;
					riscv_reg_t tdata1_test_rb;
					if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
						return ERROR_FAIL;
					if (tdata1_test_rb == tdata1_test) {
						/* Timing "before" is supported. Therefore this trigger was not hit or was hit after and single step is not needed. */
						r->need_single_step = false;
					}
				} else if (trigger_retired_info != CSR_MCONTROL6_HIT0_BEFORE) {
					/* Trigger has triggered after the instruction. */
					r->need_single_step = false;
				}
```

Reply:
the code simplify leading to below scenario missed：
trigger0 is supported `mcontrol.timing=1`, `r->need_single_step = true`;
trigger1 is supported `mcontrol6.hit0[1] = 3`, `r->need_single_step = false`;
trigger1 `r->need_single_step` info will overwrite trigger0 `r->need_single_step` .

Reply:
The issue you have mentioned can be mitigated by changing
```
r->need_single_step = false
```
to
```
r->need_single_step |= false
```

Reply:
I thank `r->need_single_step |= false` cannot be avoided muilt type trigger issue, below patch(https://github.com/riscv-collab/riscv-openocd/pull/1144/commits/5f65be97a519cc9e55acda8d8a783265e85d054b) function `mcontrol6_hit_supported` soving overwrite phenomenon, Please code review~